pkg/handler/enqueue_mapped.go

```diff
@@ -40,44 +44,58 @@ type MapFunc func(context.Context, client.Object) []reconcile.Request
 // For UpdateEvents which contain both a new and old object, the transformation function is run on both
 // objects and both sets of Requests are enqueue.
 func EnqueueRequestsFromMapFunc(fn MapFunc) EventHandler {
-	return &enqueueRequestsFromMapFunc{
+	return EnqueueRequestsFromTypedMapFunc(fn)
+}
+
+// EnqueueRequestsFromTypedMapFunc enqueues Requests by running a transformation function that outputs a collection
+// of reconcile.Requests on each Event.  The reconcile.Requests may be for an arbitrary set of objects
+// defined by some user specified transformation of the source Event.  (e.g. trigger Reconciler for a set of objects
+// in response to a cluster resize event caused by adding or deleting a Node)
+//
+// EnqueueRequestsFromTypedMapFunc is frequently used to fan-out updates from one object to one or more other
+// objects of a differing type.
+//
+// For TypedUpdateEvents which contain both a new and old object, the transformation function is run on both
+// objects and both sets of Requests are enqueue.
+func EnqueueRequestsFromTypedMapFunc[T any](fn TypedMapFunc[T]) TypedEventHandler[T] {
```

Comment:
Here is one problem as I can see. There should be an adapter method for `EventHandler` for a non-generic controller using type specific generic map method. Something like:
```go
func EnqueueRequestsFromTypedMapFunc[T client.Object](fn TypedMapFunc[T]) EventHandler
```
The problem now is that `EventHandler = TypedEventHandler[client.Object]` and you can’t type cast from one generic type to another, regardless of the fact that both implement `client.Object`. That’s where the simpler form of type aliasing broke for me, and where this (obviously simpler) design is a breaking change. Sometimes you can’t pair generics with a project due to reasons like the one we are facing - using builder or any sort of nested structures. Go generics just don’t work where there is a type ambiguity. But using generics should not be either/or decision either, it just needs a way to make a split at some point where it no longer works or does not make sense, I think.

As for real world example where this is important - I rebased CAPI testing PR on your fork + new builder methods, all works well except one common [method](https://github.com/kubernetes-sigs/cluster-api/pull/10464/files#diff-ef84e95c01630a75038b4d2cda0709d9d1d3d556459304fb39789913fb98eeaeR332). It is the only blocker for it to compile and work, but it is a deal breaker if this has to be continued (cc @sbueringer):
```go
return r.Tracker.Watch(ctx, remote.WatchInput{
    Name:         "machinepool-watchNodes",
 	Cluster:      util.ObjectKey(cluster),
 	Watcher:      &controller.ControllerAdapter{Controller: r.controller},
 	Kind:         &corev1.Node{},
 	EventHandler: handler.EnqueueRequestsFromMapFunc(r.nodeToMachinePool), // Does not compile because expects TypedMapFunc[client.Object] and gets TypedMapFunc[*corev1.Node]
 })

func (r *MachinePoolReconciler) nodeToMachinePool(ctx context.Context, node *corev1.Node) []reconcile.Request {
```
And there is no workaround except for having `TypedEventHandler` and `EventHandler` being separate interfaces with different methods working on `client.Object` and `T` simultaneously. Here is [how](https://github.com/kubernetes-sigs/controller-runtime/pull/2784/files#diff-37a34307ac3b7d0c687b0b2fd11f1a595a7987c90fc3f3b85881c3fdddfae981) I did it originally. I didn’t see a point in having `TypedEvent` structure either, as a simplification. Another user story for this - one less `controller-runtime` dependency for reusable business logic around event mapping exported in a package at the API level.

Reply:
Cluster-API was also bumped here and it doesn't look like the problem you are describing happened there: https://github.com/kubernetes-sigs/cluster-api/pull/10460

> The problem now is that EventHandler = TypedEventHandler[client.Object] and you can’t type cast from one generic type to another,

You can not type cast but you can use them interchangably. I still do not quite understand what this breaks that your approach doesn't. I am not very familiar with cluster-api, can you give a simpler example?

Reply:
I'll take a look

Reply:
> Cluster-API was also bumped here and it doesn't look like the problem you are describing happened there: https://github.com/kubernetes-sigs/cluster-api/pull/10460

I think this didn't happen here because Christian rebased on this PR, and not on "your fork + new builder methods"

Reply:
@Danil-Grigorev Can you provide more details how this func is broken? (and push a branch somewhere where it is broken)

Reply:
One detail that maybe is what you are referring to but I am not sure: It makes a difference if you do
```
type foo = bar[something]
```

or

```
type foo bar[something]
```

The latter means that you do not get the methods of `bar` and as a result, `foo` can not be used in place of `bar[something]`

But yeah, I also don't really understand the example given so I don't know if this might be what you mean.

Reply:
Yes, as I mentioned, I just had to stick with non generic `nodeToMachinePool` method which kept using `client.Object` and that probably it wasn’t the right example, because `nodeToMachinePool` is only used once - in the tracker setup.

My concern is about more complex scenarios, when the same method has to be used with generic implementation elsewhere. It can’t be done without an adapter or code duplication purely for type casting. Which might lead to decision against using generics due to added complexity in having use methods interchangeably.

The example I provided is good in outlining the second problem related to golang/go#49085 as in builder pattern UX changes. We can’t make `Tracker` a generic method with specialization on `*corev1.Node` for the CAPI use case to make it simple, because the `Tracker` is stored in `MachinePoolReconciler` which works with `MachinePools` - it would look strange and may block `MachinePoolReconciler` from being a generic method on `MachinePoolReconciler[*MachinePool]` for example if needed.

This is theoretical, I’m struggling to find a real world example which would explain the point better, still looking though. I think `controller-runtime` should take a couple more steps further to make the library UX better aligned with generics, despite the golang/go#49085 limitation, and get the design of how it looks in the end with possible edge scenarios covered.

Reply:
Sounds good to me. Tried this once again and this approach seems to work:
```go
func convert[T any](fn TypedMapFunc[T]) MapFunc {
	return func(ctx context.Context, obj client.Object) (req []reconcile.Request) {
		if typed, ok := obj.(T); !ok {
			return req
		} else {
			return fn(ctx, typed)
		}
	}
}

func EnqueueRequestsFromTypedMapFunc[T any](fn TypedMapFunc[T]) EventHandler {
	return &enqueueRequestsFromMapFunc[client.Object]{
		toRequests: convert(fn),
	}
}
```